File: frontend/lib/billing.ts
export type BillingCycle = "monthly" | "annual";

export type PlanPrice = {
  id: string;
  plan_id: string;
  plan_key: string | null;
  plan_name: string | null;
  key: string;
  billing_cycle: BillingCycle;
  currency: string;
  amount: string;
};

export type Subscription = {
  id: string;
  plan_key: string;
  plan_version: number;
  price_key: string;
  billing_cycle: BillingCycle;
  status: string;
  currency: string;
  unit_amount: string;
  current_period_start: string | null;
  current_period_end: string | null;
  grace_ends_at: string | null;
  cancel_at_period_end: boolean;
  pending_price_id: string | null;
};

export type BillingSummary = {
  billing_account_id: string | null;
  subscription: Subscription | null;
  commercial_status: string;
};

export type CheckoutSession = { id: string; status: string; provider: string; checkout_url: string | null; expires_at: string };

export type BillingInvoice = {
  id: string;
  invoice_number: string;
  status: string;
  currency: string;
  total: string;
  amount_paid: string;
  amount_due: string;
  period_start: string | null;
  period_end: string | null;
  paid_at: string | null;
  created_at: string;
  lines: { description: string; quantity: number; unit_amount: string; amount: string }[];
};

export function formatMoney(amount: string | number, currency: string) {
  return new Intl.NumberFormat(undefined, { style: "currency", currency, maximumFractionDigits: 2 }).format(Number(amount));
}

export function billingStatusMessage(summary: BillingSummary | null) {
  const subscription = summary?.subscription;
  if (!subscription) return "Your Store is using its trial or legacy commercial assignment.";
  if (subscription.status === "past_due") return subscription.grace_ends_at ? `Payment is past due. Grace access continues until ${new Date(subscription.grace_ends_at).toLocaleDateString()}.` : "Payment is past due.";
  if (subscription.cancel_at_period_end) return `Cancellation is scheduled. Access remains active until ${subscription.current_period_end ? new Date(subscription.current_period_end).toLocaleDateString() : "the paid period ends"}.`;
  return `${subscription.plan_key} is ${subscription.status}.`;
}

export function annualSavings(monthly: PlanPrice | undefined, annual: PlanPrice | undefined) {
  if (!monthly || !annual) return 0;
  return Math.max(0, Number(monthly.amount) * 12 - Number(annual.amount));
}

File: backend/app/schemas/billing.py
from datetime import datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class PlanPriceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    plan_id: UUID
    plan_key: str | None = None
    plan_name: str | None = None
    key: str
    billing_cycle: str
    currency: str
    amount: Decimal


class CheckoutCreate(BaseModel):
    plan_price_id: UUID
    idempotency_key: str = Field(min_length=8, max_length=128)


class CheckoutRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    status: str
    provider: str
    checkout_url: str | None
    expires_at: datetime


class SubscriptionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    plan_key: str
    plan_version: int
    price_key: str
    billing_cycle: str
    status: str
    currency: str
    unit_amount: Decimal
    current_period_start: datetime | None
    current_period_end: datetime | None
    grace_ends_at: datetime | None
    cancel_at_period_end: bool
    pending_price_id: UUID | None


class BillingSummaryRead(BaseModel):
    billing_account_id: UUID | None
    subscription: SubscriptionRead | None
    commercial_status: str


class InvoiceLineRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    description: str
    quantity: int
    unit_amount: Decimal
    amount: Decimal


class InvoiceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    invoice_number: str
    status: str
    currency: str
    subtotal: Decimal
    discount_total: Decimal
    tax_total: Decimal
    total: Decimal
    amount_paid: Decimal
    amount_due: Decimal
    period_start: datetime | None
    period_end: datetime | None
    paid_at: datetime | None
    created_at: datetime
    lines: list[InvoiceLineRead] = Field(default_factory=list)


class PaymentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    amount: Decimal
    currency: str
    status: str
    paid_at: datetime | None
    refunded_amount: Decimal
    created_at: datetime


class PlanChangeInput(BaseModel):
    plan_price_id: UUID


class CancellationInput(BaseModel):
    at_period_end: bool = True


class TestEventInput(BaseModel):
    outcome: Literal["success", "failed", "renewal", "refund"] = "success"


class RefundInput(BaseModel):
    amount: Decimal = Field(gt=0)
    reason: str = Field(min_length=3, max_length=500)


class PlanPriceCreate(BaseModel):
    plan_id: UUID
    key: str = Field(pattern=r"^[a-z0-9][a-z0-9-]{2,119}$")
    billing_cycle: Literal["monthly", "annual"]
    currency: str = Field(min_length=3, max_length=10)
    amount: Decimal = Field(ge=0)
